qualtran/bloqs/arithmetic/sorting.py

```diff
@@ -11,118 +11,318 @@
+                    "result",
+                    BoundedQUInt(self.bitsize, self.L),
+                    shape=(self.k * 2,),
+                    side=Side.RIGHT,
+                ),
+                Register("junk", QBit(), shape=(self.num_comparisons,), side=Side.RIGHT),
+            ]
+        )
+
+    @cached_property
+    def bitsize(self) -> SymbolicInt:
+        return bit_length(self.L - 1)
+
+    @cached_property
+    def num_comparisons(self) -> SymbolicInt:
+        return self.k * (bit_length(self.k - 1) + 1)
+
+    def is_symbolic(self):
+        return is_symbolic(self.L, self.k)
+
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', xs: 'SoquetT', ys: 'SoquetT'
+    ) -> dict[str, 'SoquetT']:
+        if self.is_symbolic():
+            raise DecomposeTypeError(f"Cannot decompose symbolic {self=}")
+
+        assert isinstance(xs, np.ndarray)
+        assert isinstance(ys, np.ndarray)
+
+        k = int(self.k)
+
+        first_round_junk = []
+        for i in range(k):
+            xs[i], ys[k - 1 - i], anc = bb.add(Comparator(self.L), a=xs[i], b=ys[k - 1 - i])
+            first_round_junk.append(anc)
+
+        result = np.concatenate([xs, ys])
+        logk = int(bit_length(k - 1))
+        assert 2**logk == k
+
+        all_junks = [first_round_junk]
+        for log_offset in reversed(range(logk)):
+            result, ancs = bb.add(
+                ParallelComparators(self.L, self.k * 2, 2**log_offset), xs=result
+            )
+            all_junks.append(ancs)
+
+        junk = np.concatenate(all_junks)
+        assert len(junk) == self.num_comparisons
+
+        return {'result': result, 'junk': np.array(junk)}
+
+
+@frozen
+class BitonicSort(Bloq):
+    r"""Sort k numbers in the range {0, ..., L-1}.
+
+    Args:
+        L: Upper-bound (excluded) on the input integers.
         k: Number of integers to sort.
 
     Registers:
-        input: List of k integers we want to sort.
+        xs: List of k integers we want to sort.
 
     References:
         [Improved techniques for preparing eigenstates of fermionic Hamiltonians](https://www.nature.com/articles/s41534-018-0071-5).
         Supporting Information Sec. II.
     """
 
-    bitsize: SymbolicInt
-    k: int
+    L: SymbolicInt
+    k: SymbolicInt
+
+    def __attrs_post_init__(self):
+        k = self.k
+        if not is_symbolic(k):
+            assert not isinstance(k, sympy.Expr)
+            assert k >= 1, "length of input list must be positive"
+            # TODO support non-power-of-two input lengths
+            assert (k & (k - 1)) == 0, "length of input list must be a power of 2"
 
     @property
     def signature(self):
-        return Signature([Register("input", QAny(self.bitsize), shape=(self.k,))])
+        return Signature(
+            [
+                Register("xs", BoundedQUInt(self.bitsize, self.L), shape=(self.k,)),
+                Register("junk", QBit(), shape=(self.num_comparisons,), side=Side.RIGHT),
```

Comment:
Why do we have `junk` RIGHT registers everywhere ? Are we saving on any uncomputation? Can you also expand the docstring to explain this?

Reply:
This is the comparator circuit from the [reference](https://www.nature.com/articles/s41534-018-0071-5):
![image](https://github.com/quantumlib/Qualtran/assets/7265746/4008c9d3-dd78-4749-9f1d-ea6d0d279aa2)

I wrote the signature similar to the `And` bloq where the result bit is `RIGHT`.

Reply:
I think its worth highlighting this in the docstring -- i.e. the junk stores the result of the comparisons. This would be necessary to make the sorting bloq reversible so you can apply the unitary in reverse and get back to the original sequence of unsorted elements if needed.

Reply:
updated the docstring explaining this.